internal/pkg/cli/deploy/env.go

```diff
@@ -100,11 +139,154 @@ func NewEnvDeployer(in *NewEnvDeployerInput) (*envDeployer, error) {
 			Env:             in.Env,
 		},
 		newStackSerializer: func(in *deploy.CreateEnvironmentInput, lastForceUpdateID string, oldParams []*awscfn.Parameter) stackSerializer {
-			return stack.NewEnvConfigFromExistingStack(in, lastForceUpdateID, oldParams)
+			return cfnstack.NewEnvConfigFromExistingStack(in, lastForceUpdateID, oldParams)
+		},
+		envDescriber:      envDescriber,
+		envManagerSession: envManagerSession,
+		lbDescriber:       elbv2.New(envManagerSession),
+		newServiceStackDescriber: func(svc string) stackDescriber {
+			return stack.NewStackDescriber(cfnstack.NameForService(in.App.Name, in.Env.Name, svc), envManagerSession)
 		},
 	}, nil
 }
 
+// Validate returns an error if the environment manifest is incompatible with services and application configurations.
+func (d *envDeployer) Validate(ctx context.Context, mft *manifest.Environment, output io.Writer) error {
+	isManagedCDNEnabled := mft.CDNEnabled() && !mft.HasImportedPublicALBCerts() && d.app.Domain != ""
+	if isManagedCDNEnabled {
+		// With managed domain, if the customer isn't using `alias` the A-records are inserted in the service stack as each service domain is unique.
+		// However, when clients enable CloudFront, they would need to update all their existing records to now point to the distribution.
+		// Hence, we force users to use `alias` and let the records be written in the environment stack instead.
+		if err := d.envDescriber.ValidateCFServiceDomainAliases(); err != nil {
+			return err
+		}
+	}
+
+	if mft.CDNEnabled() && mft.CDNDoesTLSTermination() && (mft.HasImportedPublicALBCerts() || d.app.Domain != "") {
+		err := d.validateALBWorkloadsDontRedirect(ctx)
+		var redirErr *errEnvHasPublicServicesWithRedirect
+		switch {
+		case errors.As(err, &redirErr) && mft.ALBIngressRestrictedToCDN():
+			return err
+		case errors.As(err, &redirErr):
+			fmt.Fprintln(output, redirErr.Warning())
+		case err != nil:
+			return fmt.Errorf("can't enable TLS termination on CDN: %w", err)
```

Comment:
Error from private function. We can be more verbose in `validateALBWorkloadsDontRedirect`.
```suggestion
			return err
```

Reply:
i feel like it's a little weird for the function `validateALBWorkloadsDontRedirect` to embed the context that "i'm doing this to figure out CDN termination" - what do you think?

Reply:
Yeah then omitting it would be better?

Reply:
hmmm, i don't feel like it's useful information for the users to see though - otherwise the error would just be like "verify service 'svc1': http listener not found" and have no context _why_ `env deploy` even cared